File: MathApi/NativeInterop/MathService.cs
namespace MathApi.Services;

public sealed class MathService
{
    public string Engine => "MileCalc v1.0";

    public MpgResult Mpg(double miles, double gallons)
    {
        double mpg = miles / gallons;
        return new MpgResult(R(mpg), R(235.214 / mpg), miles, gallons, Engine);
    }

    public TripCostResult TripCost(double miles, double mpg, double ppg)
    {
        double gal  = miles / mpg;
        double cost = gal * ppg;
        return new TripCostResult(R(cost), R(gal), R(cost / miles), miles, mpg, ppg, Engine);
    }

    public RangeResult Range(double tank, double mpg)
    {
        double mi = tank * mpg;
        return new RangeResult(R(mi), R(mi * 1.60934), tank, mpg, Engine);
    }

    public EmissionsResult Emissions(double miles, double mpg)
    {
        double kg = (miles / mpg) * 8.887;
        return new EmissionsResult(R(kg), R(kg * 2.20462), R(kg / 21.77), miles, mpg, Engine);
    }

    public SavingsResult Savings(double currentMpg, double newMpg, double annualMiles, double ppg)
    {
        double galSaved  = (annualMiles / currentMpg) - (annualMiles / newMpg);
        double costSaved = galSaved * ppg;
        double co2Saved  = galSaved * 8.887;
        double pctImprove = (newMpg - currentMpg) / currentMpg * 100;
        return new SavingsResult(R(costSaved), R(galSaved), R(co2Saved), R(pctImprove), currentMpg, newMpg, annualMiles, ppg, Engine);
    }

    private static double R(double v) => double.IsNaN(v) || double.IsInfinity(v) ? v : Math.Round(v, 3);
}

public sealed record MpgResult(double Mpg, double L100km, double Miles, double Gallons, string ProcessedBy);
public sealed record TripCostResult(double TotalCost, double GallonsNeeded, double CostPerMile, double Miles, double Mpg, double PricePerGallon, string ProcessedBy);
public sealed record RangeResult(double RangeMiles, double RangeKm, double Tank, double Mpg, string ProcessedBy);
public sealed record EmissionsResult(double Co2Kg, double Co2Lbs, double TreesNeeded, double Miles, double Mpg, string ProcessedBy);
public sealed record SavingsResult(double AnnualSavings, double GallonsSaved, double Co2SavedKg, double MpgImprovement, double CurrentMpg, double NewMpg, double AnnualMiles, double Ppg, string ProcessedBy);

File: math-ui/src/app/services/math.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface MpgResult       { mpg: number; l100km: number; miles: number; gallons: number; processedBy: string; }
export interface TripCostResult  { totalCost: number; gallonsNeeded: number; costPerMile: number; miles: number; mpg: number; pricePerGallon: number; processedBy: string; }
export interface RangeResult     { rangeMiles: number; rangeKm: number; tank: number; mpg: number; processedBy: string; }
export interface EmissionsResult { co2Kg: number; co2Lbs: number; treesNeeded: number; miles: number; mpg: number; processedBy: string; }
export interface GasPriceResult  { price: number; live: boolean; source: string; }
export interface SavingsResult   { annualSavings: number; gallonsSaved: number; co2SavedKg: number; mpgImprovement: number; currentMpg: number; newMpg: number; annualMiles: number; ppg: number; processedBy: string; }
export interface DbHistoryEntry  { id: number; mode: string; summary: string; createdAt: string; }

@Injectable({ providedIn: 'root' })
export class MileageService {
  private readonly http = inject(HttpClient);
  private readonly api  = `${environment.apiUrl}/api/mileage`;

  calcMpg(miles: number, gallons: number): Observable<MpgResult> {
    return this.http.get<MpgResult>(`${this.api}/mpg`,
      { params: new HttpParams().set('miles', miles).set('gallons', gallons) });
  }

  calcTripCost(miles: number, mpg: number, ppg: number): Observable<TripCostResult> {
    return this.http.get<TripCostResult>(`${this.api}/trip-cost`,
      { params: new HttpParams().set('miles', miles).set('mpg', mpg).set('ppg', ppg) });
  }

  calcRange(tank: number, mpg: number): Observable<RangeResult> {
    return this.http.get<RangeResult>(`${this.api}/range`,
      { params: new HttpParams().set('tank', tank).set('mpg', mpg) });
  }

  calcEmissions(miles: number, mpg: number): Observable<EmissionsResult> {
    return this.http.get<EmissionsResult>(`${this.api}/emissions`,
      { params: new HttpParams().set('miles', miles).set('mpg', mpg) });
  }

  calcSavings(currentMpg: number, newMpg: number, annualMiles: number, ppg: number): Observable<SavingsResult> {
    return this.http.get<SavingsResult>(`${this.api}/savings`,
      { params: new HttpParams().set('currentMpg', currentMpg).set('newMpg', newMpg).set('annualMiles', annualMiles).set('ppg', ppg) });
  }

  getGasPrice(state?: string): Observable<GasPriceResult> {
    let params = new HttpParams();
    if (state) params = params.set('state', state);
    return this.http.get<GasPriceResult>(`${this.api}/gas-price`, { params });
  }

  getHistory(): Observable<DbHistoryEntry[]> {
    return this.http.get<DbHistoryEntry[]>(`${this.api}/history`);
  }

  clearHistory(): Observable<void> {
    return this.http.delete<void>(`${this.api}/history`);
  }
}
